### PR TITLE
feat: improved SWR hooks api, useMutation result no longer cached and updated to SWR 2.0

### PR DESCRIPTION
**Pull request by Pagebakers**
#### Checklist

- [x] run `make test`
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

Fixes ENG-570, ENG-595, ENG-395

New and improved docs will be added in #319

### CODE REVIEW

File: packages/swr/src/hooks.ts [hunk 1]
```diff
@@ -153,80 +203,184 @@ export const createHooks = <Operations extends OperationsDefinition>(client: Cli
...
+
+		const startedAtRef = useRef<number | null>(null);
+
+		const [state, setState] = useState({
+			isLoading: false,
+			isSubscribed: false,
+		});
+
+		useEffect(() => {
+			if (!startedAtRef.current && resetOnMount) {
```
Comment (Pagebakers) [inline packages/swr/src/hooks.ts:266]: Allow  to reset the cached data when the hook in mounted, so we don't have 'old' subscription values.

@StarpTech what do you think of the `resetOnMount` prop name?

Comment (StarpTech) [reply packages/swr/src/hooks.ts:266]: I like the name.

File: packages/sdk/src/client/client.ts [hunk 2]
```diff
@@ -247,23 +247,31 @@ export class Client {
 		subscription: SubscriptionRequestOptions,
```
Comment (StarpTech) [inline packages/sdk/src/client/client.ts:247]: Don't forget to add tests for the types https://github.com/wundergraph/wundergraph/blob/main/packages/type-tests/test/swr.test-d.ts. It's a great place to understand the API of the new client.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/swr/src/hooks.ts [hunk 3]
```diff
@@ -153,80 +203,184 @@ export const createHooks = <Operations extends OperationsDefinition>(client: Cli
...
-	>(
-		options: UseSubscriptionOptions<OperationName, Input>
-	): SWRResponse<Data, GraphQLResponseError> => {
-		const { enabled = true, operationName, input, subscribeOnce, ...swrConfig } = options;
-		const key = { operationName, input };
-		const _key = serialize(key);
-		const response = useSWR(enabled ? key : null, null, swrConfig);
+	// Helper hook used in useQuery and useSubscription
+	const useSubscribeTo = (props: UseSubscribeToProps) => {
+		const { mutate } = useSWRConfig();
```
Comment (StarpTech) [inline packages/swr/src/hooks.ts:254]: Does that mean the user has to use `swr` context?

File: packages/swr/src/hooks.ts [hunk 4]
```diff
@@ -1,38 +1,25 @@
-import useSWR, { mutate, SWRConfiguration, SWRResponse, MutatorOptions } from 'swr';
+import useSWR, { SWRConfiguration, useSWRConfig } from 'swr';
+import useSWRMutation, { SWRMutationConfiguration } from 'swr/mutation';
```
Comment (StarpTech) [inline packages/swr/src/hooks.ts:2]: Can we set up a minimal test suite? We could use swr tests as an inspiration.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: packages/swr/src/hooks.ts (hunk 3 above)
Comment (Pagebakers) [reply packages/swr/src/hooks.ts:254]: No, SWR always fallback to global context by default.

File: packages/swr/src/hooks.ts (hunk 4 above)
Comment (Pagebakers) [reply packages/swr/src/hooks.ts:2]: Yes I think it's a good idea, will look into this today.

File: packages/swr/src/hooks.ts (hunk 3 above)
Comment (StarpTech) [reply packages/swr/src/hooks.ts:254]: ok

File: packages/sdk/src/client/client.ts (hunk 2 above)
Comment (Pagebakers) [reply packages/sdk/src/client/client.ts:247]: Fixed tests

File: packages/sdk/src/client/client.ts [hunk 5]
```diff
@@ -247,23 +251,31 @@ export class Client {
...
+				wg_variables: this.stringifyInput(subscription.input),
+				wg_live: subscription?.liveQuery ? 'true' : 'false',
+				wg_sse: 'true',
+			});
+			const url = this.addUrlParams(this.operationUrl(subscription.operationName), params);
+			const eventSource = new EventSource(url, {
+				withCredentials: true,
+			});
+			eventSource.addEventListener('error', () => {
+				reject(new Error(`SSE connection error: ${subscription.operationName}`));
```
Comment (StarpTech) [inline packages/sdk/src/client/client.ts:266]: Can we log / enrich the original error? `(err)`

File: packages/swr/package.json [hunk 6]
```diff
@@ -17,17 +18,27 @@
     "url": "git+https://github.com/wundergraph/wundergraph.git"
   },
   "dependencies": {
-    "swr": "^1.3.0"
+    "swr": "^2.0.0-rc.0"
   },
   "peerDependencies": {
     "@wundergraph/sdk": ">=0.91.0",
```
Comment (StarpTech) [inline packages/swr/package.json:24]: Update it?

File: packages/swr/src/hooks.test.tsx [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	};
+};
+
+describe('SWR - createHooks', () => {
+	const client = createClient();
+
+	const hooks = createHooks(client);
+
+	it('should be able to init hooks', async () => {
+		expect(hooks).toBeTruthy();
```
Comment (StarpTech) [inline packages/swr/src/hooks.test.tsx:100]: Can we validate here if `createHooks` returns the correct api?

File: packages/swr/src/types.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+export interface SubscribeToOptions extends SubscriptionRequestOptions {
+	onResult(response: ClientResponse): void;
+	onSuccess?(response: ClientResponse): void;
+	onError?(error: GraphQLResponseError): void;
+	onAbort?(): void;
+}
+
+export interface UseUserOptions<User> extends FetchUserRequestOptions, SWRConfiguration<User> {
```
Comment (StarpTech) [inline packages/swr/src/types.ts:131]: `UseUserHookOptions` ?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: testapps/nextjs/.wundergraph/wundergraph.server.ts [hunk 9]
```diff
@@ -86,6 +86,37 @@ export default configureWunderGraphServer<HooksConfig, InternalClient>(() => ({
 				},
 			},
 		},
+		subscriptions: {
+			// .wundergraph/operations/Counter.graphql
```
Comment (StarpTech) [inline testapps/nextjs/.wundergraph/wundergraph.server.ts:90]: Delete?

**Comment by StarpTech** [comment]
Please rename the PR title to something more meaningful for the user e.g improve mutation support.

File: packages/sdk/src/client/client.ts (hunk 5 above)
Comment (Pagebakers) [reply packages/sdk/src/client/client.ts:266]: There was no valuable information available in the parameter, as it's just a generic event.

https://developer.mozilla.org/en-US/docs/Web/API/EventSource/error_event

Comment (StarpTech) [reply packages/sdk/src/client/client.ts:266]: ok

File: packages/swr/src/hooks.test.tsx (hunk 7 above)
Comment (StarpTech) [reply packages/swr/src/hooks.test.tsx:100]: ok

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**